### ISSUE TITLE
--log-file seemingly has no effect, no log files written

### ISSUE THREAD

**Issue by brsolomon-deloitte** [kubernetes/kubeadm#2455]
## Is this a BUG REPORT or FEATURE REQUEST?

BUG REPORT

## Description

Passing `--log-file` to both `kubeadm init` and `kubeadm join` never actually creates or writes a log file whatever the level of `--v` and regardless of whether the enclosing directory exists.

## Versions

**kubeadm version** : `kubeadm version: &version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:30:03Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}`

**Environment**:
- **Kubernetes version**: `Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}`
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): Ubuntu 18.04.5
- **Kernel** (e.g. `uname -a`): 5.4.0

## What happened?

```bash
$ ls -la /var/log/kubeadm/
total 8
drwxr-xr-x  2 root root   4096 Apr 21 22:48 .
drwxrwxr-x 12 root syslog 4096 Apr 21 22:51 ..
```

## What you expected to happen?

A log file to be produced at `/var/log/kubeadm/join-*.log`

## How to reproduce it (as minimally and precisely as possible)?

```bash
kubeadm join xxxx.elb.us-east-2.amazonaws.com:6443 --token xxx.xxx \
        --discovery-token-ca-cert-hash sha256:xxxx \
        --control-plane --certificate-key xxxx \
        --log-file /var/log/kubeadm/join-$(date +%s).log --one-output --v 9
```

**Comment by neolit123** [comment]
i can confirm that --log-file does not work for some reason.
...but it also does not work for a `kubelet --v 10 --log-file ./foo.log`.

both kubeadm and kubelet use klog and these flags come from it:
https://github.com/kubernetes/klog

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/kubeadm.go#L33-L49
so there must be some sort of a klog bug at play here.

can you please log a kubernetes/kubernets ticket and tag it with `/sig instrumentation` & `/kind bug`.

**Comment by brsolomon** [comment]
> can you please log a kubernetes/kubernetes ticket and tag it with /sig instrumentation & /kind bug.

I am over my head in work at the moment, I'll do that when I can. For now I have resorted to `| tee <log file>`

**Comment by neolit123** [comment]
actually never mind...no need to log an issue there.
the problem is that there you also need to pass `--logtostderr=false`

this is done for backwards compatibility IIRC.
you need to pass a log file but also disable log to the standard err stream.

your report opens up for a couple of problems in kubeadm:

- i don't know why it hides the `logtostderr` flag:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/kubeadm.go#L37-L47
maybe it was deprecated, but if so how can one log to a file in the first place?

- one cannot mix `--config` with `--log-file` which doesn't make sense.
basically all klog flags are blocked currently:
https://github.com/kubernetes/kubernetes/blob/89db31aadefb5207dbf84e2403495d0da31c7c35/cmd/kubeadm/app/apis/kubeadm/validation/validation.go#L539-L571
and maybe we need to allow `logtostderr` if its not deprecated.

**Comment by neolit123** [comment]
^ adding the help label if someone wishes to help us with investigation here.
/help

**Comment by neolit123** [comment]
@brsolomon and you current workaround is not not use the log-file flag at all and just pipe the kubeadm stderr and/or stdout output to a file.

**Comment by brsolomon** [comment]
> @brsolomon and you current workaround is not not use the log-file flag at all and just pipe the kubeadm stderr and/or stdout output to a file.

Yep. Just dropping `--log-file` altogether and

```bash
kubeadm command <args> | tee /var/log/kubeadm/command-$(date +%s).log
```

**Comment by brsolomon** [comment]
Yes, oddly enough, confirmed that you _must_ pass `--logtostderr=false` in conjunction with `--log-file` for `--log-file` to actually log to the intentded file.

**Comment by lojies** [comment]
[#64974](https://github.com/kubernetes/kubernetes/pull/64974) marked these flags as hidden.
@luxas, can you give some suggestion?

**Comment by CaoDonghui123** [comment]
> Yes, oddly enough, confirmed that you _must_ pass `--logtostderr=false` in conjunction with `--log-file` for `--log-file` to actually log to the intentded file.

there  code set ` --logtostderr=true`  means "log to standard error instead of files" 
https://github.com/kubernetes/kubernetes/blob/113f828553462a6e086a2bd0ba044c5ae20f61a5/cmd/kubeadm/app/kubeadm.go#L37
